[mlir][vector] Group re-order patterns together

banach-space opened the pull request:
Group all patterns that re-order vector.transpose and vector.broadcast
Ops (*) under `populateSinkVectorOpsPatterns`. These patterns are
normally used to "sink" redundant Vector Ops, hence grouping together.
Example:

```mlir
%at = vector.transpose %a, [1, 0]: vector<4x2xf32> to vector<2x4xf32>
%bt = vector.transpose %b, [1, 0]: vector<4x2xf32> to vector<2x4xf32>
%r = arith.addf %at, %bt : vector<2x4xf32>
```
would get converted to:
```mlir
%0 = arith.addf %a, %b : vector<4x2xf32>
%r = vector.transpose %0, [1, 0] : vector<2x4xf32>
```

This patch also moves all tests for these patterns so that all of them
are:
  * run under one test-flag: `test-vector-sink-patterns`,
  * located in one file: "vector-sink.mlir".

To facilitate this change:
  * `-test-sink-vector-broadcast` is renamed as
    `test-vector-sink-patterns`,
  * "sink-vector-broadcast.mlir" is renamed as "vector-sink.mlir",
  * tests for `ReorderCastOpsOnBroadcast` and
    `ReorderElementwiseOpsOnTranspose` patterns are moved from
    "vector-reduce-to-contract.mlir" to "vector-sink.mlir",
  * `ReorderElementwiseOpsOnTranspose` patterns are removed from
    `populateVectorReductionToContractPatterns` and added to (newly
    created) `populateSinkVectorOpsPatterns`,
  * `ReorderCastOpsOnBroadcast` patterns are removed from
    `populateVectorReductionToContractPatterns` - these are already
    present in `populateSinkVectorOpsPatterns`.

This should allow us better layering and more straightforward testing.
For the latter, the goal is to be able to easily identify which pattern
a particular test is exercising (especially when it's a specific
pattern).

NOTES FOR DOWNSTREAM USERS

In order to preserve the current functionality, please make sure to add
  * `populateSinkVectorOpsPatterns`,

wherever you are using `populateVectorReductionToContractPatterns`.
Also, rename `populateSinkVectorBroadcastPatterns` as
`populateSinkVectorOpsPatterns`.

(*) I didn't notice any other re-order patterns.

banach-space committed:
[mlir][vector] Group tests for re-order patterns

Moves all tests for patterns that re-order vector.transpose and
vector.broadcast Ops (*) under one test-flag:
  * `test-vector-reorder-patterns`.

To facilitate this,
  * `-test-sink-vector-broadcast` is renamed as
    `test-vector-reorder-patterns`,
  * "sink-vector-broadcast.mlir" is renamed as "vector-reorder.mlir",
  * tests for `ReorderCastOpsOnBroadcast` and
    `ReorderElementwiseOpsOnTranspose` patterns are moved from
    "vector-reduce-to-contract.mlir" to "vector-reorder.mlir",
  * `ReorderElementwiseOpsOnTranspose` patterns are removed from
    `populateVectorReductionToContractPatterns` and added to (newly
    created) `populateReoderVectorTransposePatterns`.
  * `ReorderCastOpsOnBroadcast` patterns are removed from
    `populateVectorReductionToContractPatterns` - these are already
    present in `populateSinkVectorBroadcastPatter`.

This should allow us better layering and more straightforward testing.
For the latter, the goal is to be able to easily identify which pattern
a particular test is exercising (especially when it's a specific
pattern).

Note for downstream users: in order to preserve the current
functionality, please make sure to add
  * `populateReoderVectorTransposePatterns` and
    `populateSinkVectorBroadcastPatter`,

wherever you are using `populateVectorReductionToContractPatterns`

(*) I didn't notice any other re-order patterns.

MacDue commented on mlir/test/lib/Dialect/Vector/TestVectorTransforms.cpp:
```diff
@@ -385,16 +385,17 @@ struct TestSinkVectorBroadcast
   }
 
-  StringRef getArgument() const final { return "test-sink-vector-broadcast"; }
+  StringRef getArgument() const final { return "test-vector-reorder-patterns"; }
```
You've renamed the argument but the class is still `TestSinkVectorBroadcast`.

banach-space replied:
Good catch, updated!

banach-space committed:
fixup! [mlir][vector] Group tests for re-order patterns

Rename TestSinkVectorBroadcast as TestVectorReorderPatterns

MacDue commented on mlir/include/mlir/Dialect/Vector/Transforms/VectorRewritePatterns.h:
```diff
@@ -148,6 +148,10 @@ void populateVectorTransferCollapseInnerMostContiguousDimsPatterns(
 void populateSinkVectorBroadcastPatterns(RewritePatternSet &patterns,
                                          PatternBenefit benefit = 1);
 
+/// Patterns that re-order transpose Ops.
+void populateReoderVectorTransposePatterns(RewritePatternSet &patterns,
+                                           PatternBenefit benefit = 1);
```
It would be good to explain in the the docs _why_ you may want to use these patterns. This line doc comment does not give me any more information than the function name.

MacDue commented on mlir/lib/Dialect/Vector/Transforms/VectorTransforms.cpp:
```diff
@@ -2030,8 +2030,7 @@ void mlir::vector::populateVectorContractCanonicalizeMatmulToMMT(
 void mlir::vector::populateVectorReductionToContractPatterns(
     RewritePatternSet &patterns, PatternBenefit benefit) {
   patterns.add<MultiReduceToContract, CombineContractBroadcast,
-               CombineContractABTranspose, CombineContractResultTranspose,
-               ReorderCastOpsOnBroadcast, ReorderElementwiseOpsOnTranspose>(
```
Note: `populateVectorReductionToContractPatterns()` is used a few times in IREE, so this change will likely cause some breakages there. You should probably share a fix on the Discord before landing this :slightly_smiling_face:

MaheshRavishankar replied:
Thanks for the heads up! So is the expectation that you run the sink before you run contract conversion patterns.

I agree the current grouping of patterns is adhoc, but what is the expected path for users (iree and others) to get back the same code that this one method would give?

banach-space replied:
>  So is the expectation that you run the sink before you run contract conversion patterns.

See the summary :) 

> NOTES FOR DOWNSTREAM USERS
>
> In order to preserve the current functionality, please make sure to add
> 
> * populateSinkVectorOpsPatterns,
> 
> wherever you are using populateVectorReductionToContractPatterns.
> Also, rename populateSinkVectorBroadcastPatterns as
> populateSinkVectorOpsPatterns.

I checked MLIR and IREE and in both cases these were required _after_. In IREE, I run `ctest -R Codegen/SPIRV/` and `ctest -R Codegen/LLVMCPU` and both pass 100%. This is my diff:
```diff
diff --git a/compiler/src/iree/compiler/Codegen/Common/EmulateNarrowType.cpp b/compiler/src/iree/compiler/Codegen/Common/EmulateNarrowType.cpp
index 1824fb08bb..e2c8805d27 100644
--- a/compiler/src/iree/compiler/Codegen/Common/EmulateNarrowType.cpp
+++ b/compiler/src/iree/compiler/Codegen/Common/EmulateNarrowType.cpp
@@ -146,7 +146,7 @@ struct EmulateNarrowTypePass final
     }

     RewritePatternSet sinkBroadcast(ctx);
-    vector::populateSinkVectorBroadcastPatterns(sinkBroadcast);
+    vector::populateSinkVectorOpsPatterns(sinkBroadcast);
     if (failed(applyPatternsAndFoldGreedily(getOperation(),
                                             std::move(sinkBroadcast)))) {
       getOperation()->emitOpError("failed in sinking of broadcasts");
diff --git a/compiler/src/iree/compiler/Codegen/Common/GenericVectorization.cpp b/compiler/src/iree/compiler/Codegen/Common/GenericVectorization.cpp
index 98814d1342..8aee5ba2c0 100644
--- a/compiler/src/iree/compiler/Codegen/Common/GenericVectorization.cpp
+++ b/compiler/src/iree/compiler/Codegen/Common/GenericVectorization.cpp
@@ -409,6 +409,7 @@ void GenericVectorizationPass::runOnOperation() {
     vector::populateVectorTransferPermutationMapLoweringPatterns(
         vectorizationPatterns);
     vector::populateVectorReductionToContractPatterns(vectorizationPatterns);
+    vector::populateSinkVectorOpsPatterns(vectorizationPatterns);
   }
   if (foldCastIntoContract) {
     vector::populateFoldArithExtensionPatterns(vectorizationPatterns);
diff --git a/compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp b/compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp
index aa72280e5f..f00af42845 100644
--- a/compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp
+++ b/compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorCoreVectorization.cpp
@@ -96,6 +96,7 @@ public:
       vector::populateVectorTransferPermutationMapLoweringPatterns(
           contractionPatterns);
       vector::populateVectorReductionToContractPatterns(contractionPatterns);
+      vector::populateSinkVectorOpsPatterns(contractionPatterns);
       if (failed(applyPatternsAndFoldGreedily(
               funcOp, std::move(contractionPatterns)))) {
         return signalPassFailure();
diff --git a/compiler/src/iree/compiler/Codegen/SPIRV/SPIRVInitialVectorLowering.cpp b/compiler/src/iree/compiler/Codegen/SPIRV/SPIRVInitialVectorLowering.cpp
index 493894675c..c24da49689 100644
--- a/compiler/src/iree/compiler/Codegen/SPIRV/SPIRVInitialVectorLowering.cpp
+++ b/compiler/src/iree/compiler/Codegen/SPIRV/SPIRVInitialVectorLowering.cpp
@@ -314,6 +314,7 @@ public:
       // cancel them or embed into contract ops. Embedding in the flexible
       // contract ops will help to sustain the structure through various
       // transformations.
+      vector::populateSinkVectorOpsPatterns(patterns);
       vector::populateVectorReductionToContractPatterns(patterns);
       // Pull in patterns to canonicalize transfer ops.
       vector::populateVectorTransferPermutationMapLoweringPatterns(patterns);
diff --git a/third_party/llvm-project b/third_party/llvm-project
index 6b7afaa9db..dc678f1dfe 160000
--- a/third_party/llvm-project
+++ b/third_party/llvm-project
@@ -1 +1 @@
-Subproject commit 6b7afaa9db8f904ebf0262774e38e54b36598782
+Subproject commit dc678f1dfe49cd4d0b9136ff2490a482ae91e786
```

@MaheshRavishankar anything else that I should check?

Btw, this re-grouping has been a bit tricky to verify 100% - there are no tests that would require these patterns to be run together (otherwise I wouldn't be able to move the tests around as I did).

MaheshRavishankar replied:
Thanks so much for doing this! Part of my concern here as well was that this combination of things wasnt tested properly in tree. Not your fault, it is just how this was done (there are implicit assumptions of what patterns go together that isnt always clear in the vector dialect).

MacDue commented on mlir/lib/Dialect/Vector/Transforms/VectorTransforms.cpp:
```diff
@@ -2049,6 +2048,12 @@ void mlir::vector::populateSinkVectorBroadcastPatterns(
       patterns.getContext(), benefit);
 }
 
+void mlir::vector::populateReoderVectorTransposePatterns(
```
Above: should `populateSinkVectorBroadcastPatterns()` be `populateReorderVectorBroadcastPatterns()` for consistency?

MacDue commented on mlir/lib/Dialect/Vector/Transforms/VectorTransforms.cpp:
```diff
@@ -2049,6 +2048,12 @@ void mlir::vector::populateSinkVectorBroadcastPatterns(
 }
 
+void mlir::vector::populateReoderVectorTransposePatterns(
+    RewritePatternSet &patterns, PatternBenefit benefit) {
+  patterns.add<ReorderElementwiseOpsOnTranspose>(patterns.getContext(),
+                                                 benefit);
+}
```
This only adds a single pattern. Should `populateSinkVectorBroadcastPatterns()` and `populateReoderVectorTransposePatterns()` be separate things (they're not used separately here).

Why not just `populateVectorReoderPatterns()`? 
If you need more control you could still add options to the populate call.

banach-space committed:
fixup! fixup! [mlir][vector] Group tests for re-order patterns

Group all re-order patterns under `populateSinkVectorOpsPatterns`.
Rename test flag and test pass accordingly.

banach-space commented:
@MacDue Thanks for the feedback! Having thought about it, I feel that grouping all "reorder"/"sink" patterns under one `populate_Patterns()` hook would make most sense. I decided to name it as `populateVectorSinkPatterns()` - it nicely conveys _why_ that you suggested documenting.

I've renamed all the relevant hooks, passes and flags. If you agree that this is the right direction, I'll update the summary and make this PR about patterns rather than tests (grouping tests would be a nice side-effect). At which point, I'll ping a downstream user (e.g. from IREE) to also approve this. How does it sound?

Btw, I think that the updated version addresses your comments?

MacDue commented on mlir/test/Dialect/Vector/vector-reorder.mlir:
```diff
@@ -0,0 +1,321 @@
+// RUN: mlir-opt %s -test-vector-sink-patterns -split-input-file | FileCheck %s
```
Should this test file be called `vector-sink.mlir?` (matching the pass name)

MacDue reviewed (approved):
The updated patch LGTM now :+1: Just one small thing:

banach-space committed:
fixup! fixup! fixup! [mlir][vector] Group tests for re-order patterns

Rename test

MaheshRavishankar reviewed (changes requested):
Thanks for flagging this to us. Just one question

nujaa commented on mlir/test/Dialect/Vector/vector-sink.mlir:
```diff
@@ -208,3 +208,115 @@ func.func @negative_op_only_supports_vectors(%arg0 : f32) -> vector<1xf32> {
   %1 = vector.fma %0, %0, %0 : vector<1xf32>
   return %1 : vector<1xf32>
 }
+
```
Diff-ed moved tests. Looks good. 👍

nujaa reviewed (approved):
This makes lots of sense, thanks. It is definitely going to break lots of things downstream, indeed.

nujaa commented:
> Thanks for flagging this to us. Just one question

I think this has been addressed in the commit description under
`NOTES FOR DOWNSTREAM USERS`

MaheshRavishankar reviewed (approved):
Unblocking this, but I am trying this out downstream based on the extremely helpful changes from Andrej

banach-space commented:
> Unblocking this, but I am trying this out downstream based on the extremely helpful changes from Andrej

Thanks! I'm going to land it now, but please just revert if there  are any issues. I will be away next week, so won't be able to revert myself.

Thank you all for reviewing 🙏🏻
